guardian/support-dotcom-components
# Two-tier geographical targeting for amounts tests

## What does this change?
Marketing have asked us to supply them with the ability to set up a contribution amounts test for a given country, in addition to the tests for existing regions.

+ Note that this request represents a major refactor of the data model used to store amounts tests, which previously restricted amounts test data to just our 7 named regions.

Associated PRs in other repos
+ SAC: https://github.com/guardian/support-admin-console/pull/439
+ SDC: https://github.com/guardian/support-dotcom-components/pull/898
+ Support Frontend: https://github.com/guardian/support-frontend/pull/5027

This PR updates the Choice cards component in the Epic, and certain Banner templates, to use the new V2 amounts model data. More importantly it makes changes on the server side to first check to see if an amounts test exists for their reported country code (eg: `FR` for France) and only falls back to the applicable regional amounts test if no country-specific amounts test can be found in the data.

Affected components:
+ `ContributionsEpic`
+ `ChoiceCardsBanner`
+ `ChoiceCardsButtonsBanner`
+ `MomentTemplateBanner`
+ `WorldPressFreedomDayBanner`

#### To be aware!
The new amounts data model includes new attribute for each test variant:
+ `defaultContributionType` - this used to be configured in RRCP in the `Contributions Types` page, but more recently the value has been settable when building Epic message tests in RRCP. For banners with choice cards, the value has been hard-coded (setting monthly contributions as the default). The RRCP `Contributions Types` page and the Epic functionality are now deprecated; going forward this attribute will only be settable in amounts test variants.
+ `displayContributionType` - determine whether a contribution tab will be displayed in the Choice card. Used to be configured in RRCP in the `Contributions Types` page, but has been long ignored/deprecated. going forward this attribute will only be settable in amounts test variants. 

Note that this PR updates Epics to respect the settings of the `displayContributionType` attribute; Banner choice cards have not yet been updated!

## Navigating through the PR
This PR touches 31 files, as follows:

**/cdk/lib folder** (2 files)
Covers required tooling changes to give the backend access to the new Amounts V2 JSON file stored in AWS S3

**/packages/modules/src folders** (22 files)
These are fairly regular changes across components that use Choice Cards, to make them work with the new Amounts V2 data model.

**/packages/server/src folders** (5 files)
These changes are critical, as this is where the server determines which Amounts test variant is displayed in an amounts card to the user, dependent on that user's reported locality. The AB test variant selection code mirrors the code used in support-frontend.

The following routers are affected:
+ `ampEpicRouter` (which continues to show default amounts)
+ `bannerRouter`
+ `epicRouter`

**/packages/shared/src folders** (2 files)
The new Amounts V2 model is typed in the `types/abTests/epic.ts` file, while the necessary changes to determine if an Amounts test's `target` value is a valid country code, or region code, is handled in the `lib/geolocations.ts` file.

## How to test
Testing is in 3 parts:
+ Storybook
+ End-to-end testing in CODE to make sure that a country-specific amounts test displays correctly in Epics and Banners for a user in a country with an associated amounts test, that details of the correct amounts test get coded into the support site URL, and that clicking on the CTA will take the user to the support site which displays the same amounts data and choice as the user saw/made in the Epic/Banner. [Link to testing document](https://docs.google.com/document/d/1P-MS1Do_5Xqro6E6KKiw_7gID8c4U3L_CE_nAn-j2VI/edit).
+ A check to make sure the correct data is logged in the correct places and can be found using the normal data analysis tools.

## Screenshots
This PR does not change the design of the various choice cards in the Epic or Banners, except:
+ The Epic should react appropriately to the values in the amounts data's `displayContributionType` attribute

![Screenshot 2023-05-18 at 14 50 49](https://github.com/guardian/support-dotcom-components/assets/5357530/39640dd3-d8ee-47e0-9f55-79f2bccaa7de)

<img width="605" alt="Screenshot 2023-05-18 at 14 51 50" src="https://github.com/guardian/support-dotcom-components/assets/5357530/0d8b6c73-dfd8-473e-ae8e-6b03b31e9476">

![Screenshot 2023-05-18 at 14 52 10](https://github.com/guardian/support-dotcom-components/assets/5357530/1e7327ab-da1c-4511-96cc-caaf8b1ab644)
